### ISSUE TITLE
Kernel builds don't properly track dependencies between 'build' and 'push' targets

### ISSUE THREAD

**Issue by kojustin** [linuxkit/linuxkit#2994]
**Description**
The kernel makefile doesn't properly do tracking of build products between the "build_X" and "push_X" targets. Following the instructions in https://github.com/linuxkit/linuxkit/blob/master/docs/kernels.md I built a kernel like this.

`make -C kernel ORG=$MY_COOL_ORG NOTRUST=1 build_4.14.x`

the result is that you get a Docker image of the kernel in the local docker daemon. After some testing I decided that I like this kernel and I attempt to push it to my registry.

`make -C kernel ORG=$MY_COOL_ORG NOTRUST=1 push_4.14.x`

The `push_X` targets have as a prerequisite the corresponding `build_X` target. 

The `build_X` checks to see if it needs to do work by performing a `docker pull` of the kernel name. Since my original make invocation was a local build the pull will fail because the registry won't have my kernel yet. The check for an appropriate kernel fails so the second make invocation will perform a complete rebuild of the kernel.

**Comment by justincormack** [comment]
Its always difficult handling docker images in makefiles...

**Comment by kojustin** [comment]
Agreed.

**Comment by ijc** [comment]
Does it really do a full rebuild or are all the stages coming from the cache the second time round?

**Comment by kojustin** [comment]
@ijc good point. Yeah I was confused about why wouldn't reuse the cache. Turns out all calls to "docker build" in the kernel Makefile pass the "--no-cache" option. I went back through history and it looks like it's been this way since 43d981094d.
